File: app/src/main/java/com/smartwalkie/voiceping/LoginActivity.java
package com.smartwalkie.voiceping;

import android.Manifest;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.smartwalkie.voicepingsdk.callbacks.ConnectCallback;
import com.smartwalkie.voicepingsdk.exceptions.PingException;

import java.util.List;

import pub.devrel.easypermissions.EasyPermissions;

public class LoginActivity extends AppCompatActivity implements
        EasyPermissions.PermissionCallbacks {

    public static final String TAG = LoginActivity.class.getSimpleName();

    private final int RC_RECORD_AUDIO = 1000;

    private EditText userIdText;
    private View progressView;
    private View connectFormView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        userIdText = (EditText) findViewById(R.id.user_id_text);

        Button connectButton = (Button) findViewById(R.id.connect_button);
        connectButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                EasyPermissions.requestPermissions(LoginActivity.this,
                        "This app needs your permission to allow recording audio",
                        RC_RECORD_AUDIO,
                        Manifest.permission.RECORD_AUDIO);
            }
        });

        connectFormView = findViewById(R.id.connect_form);
        progressView = findViewById(R.id.connect_progress);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this);
    }

    @Override
    public void onPermissionsGranted(int requestCode, List<String> perms) {
        if (requestCode == RC_RECORD_AUDIO) attemptToLogin();
    }

    @Override
    public void onPermissionsDenied(int requestCode, List<String> perms) {
        Toast.makeText(this, "You need to allow the permission request!", Toast.LENGTH_SHORT).show();
    }

    /**
     * Attempts to connect using details specified by the form.
     * If there are form errors, the errors are presented and no actual login attempt is made.
     */
    private void attemptToLogin() {
        // Reset errors.
        userIdText.setError(null);

        // Store values at the time of the connect attempt.
        final String userId = userIdText.getText().toString().trim();

        // Check for a valid username.
        if (userId == null || userId.isEmpty()) {
            userIdText.setError(getString(R.string.error_invalid_user_id));
            userIdText.requestFocus();
        } else {
            // Show a progress spinner, and kick off a background task to
            // perform the user connect attempt.
            showProgress(true);
            VoicePingClientApp.getVoicePing().connect(userId, new ConnectCallback() {
                @Override
                public void onConnected() {
                    Log.v(TAG, "onConnected");
                    showProgress(false);
                    Intent intent = new Intent(LoginActivity.this, MainActivity.class);
                    intent.putExtra("user_id", userId);
                    startActivity(intent);
                    finish();
                }

                @Override
                public void onFailed(final PingException exception) {
                    Log.v(TAG, "onFailed");
                    showProgress(false);
                    Toast.makeText(LoginActivity.this, exception.getMessage(), Toast.LENGTH_SHORT)
                            .show();
                }
            });
        }
    }

    private void showProgress(boolean show) {
        progressView.setVisibility(show ? View.VISIBLE : View.GONE);
        connectFormView.setVisibility(show ? View.GONE : View.VISIBLE);
    }
}

File: app/src/main/java/com/smartwalkie/voicepingdemo/LoginActivity.kt
package com.smartwalkie.voicepingdemo

import android.Manifest
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import pub.devrel.easypermissions.EasyPermissions.PermissionCallbacks
import android.os.Bundle
import pub.devrel.easypermissions.EasyPermissions
import android.widget.Toast
import com.smartwalkie.voicepingsdk.callback.ConnectCallback
import android.util.Log
import android.view.View
import com.smartwalkie.voicepingdemo.databinding.ActivityLoginBinding
import com.smartwalkie.voicepingsdk.VoicePing
import com.smartwalkie.voicepingsdk.exception.VoicePingException

class LoginActivity : AppCompatActivity(), PermissionCallbacks {
    private val TAG = "LoginActivity"
    private val RC_RECORD_AUDIO = 1000
    private lateinit var binding: ActivityLoginBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)
        supportActionBar?.title = getString(R.string.app_name)
        binding.editServerUrl.setText(MyPrefs.serverUrl ?: "")
        binding.buttonConnect.setOnClickListener {
            EasyPermissions.requestPermissions(
                this@LoginActivity,
                "This app needs your permission to allow recording audio",
                RC_RECORD_AUDIO,
                Manifest.permission.RECORD_AUDIO
            )
        }
    }

    override fun onStart() {
        super.onStart()
        val userId = MyPrefs.userId ?: ""
        val company = MyPrefs.company ?: ""
        val serverUrl = MyPrefs.serverUrl ?: ""
        if (userId.isNotBlank() && company.isNotBlank() && serverUrl.isNotBlank()) {
            startActivity(Intent(this, MainActivity::class.java))
            finish()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this)
    }

    override fun onPermissionsGranted(requestCode: Int, perms: List<String>) {
        if (requestCode == RC_RECORD_AUDIO) attemptToLogin()
    }

    override fun onPermissionsDenied(requestCode: Int, perms: List<String>) {
        Toast.makeText(this, "You need to allow the permission request!", Toast.LENGTH_SHORT).show()
    }

    /**
     * Attempts to connect using details specified by the form.
     * If there are form errors, the errors are presented and no actual login attempt is made.
     */
    private fun attemptToLogin() {
        // Reset errors.
        binding.editUserId.error = null
        binding.editCompany.error = null
        binding.editServerUrl.error = null

        // Store values at the time of the connect attempt.
        val userId = binding.editUserId.text.toString().trim { it <= ' ' }
        val company = binding.editCompany.text.toString().trim { it <= ' ' }
        val serverUrl = binding.editServerUrl.text.toString().trim { it <= ' ' }

        when {
            userId.isEmpty() -> {
                binding.editUserId.error = getString(R.string.cannot_be_blank)
                binding.editUserId.requestFocus()
            }
            company.isEmpty() -> {
                binding.editCompany.error = getString(R.string.cannot_be_blank)
                binding.editCompany.requestFocus()
            }
            serverUrl.isEmpty() -> {
                binding.editServerUrl.error = getString(R.string.cannot_be_blank)
                binding.editServerUrl.requestFocus()
            }
            else -> {
                // Show a progress spinner, and kick off a background task to perform the user connect attempt.
                Utils.closeKeyboard(this, currentFocus)
                showProgress(true)
                VoicePing.connect(serverUrl, userId, company, object : ConnectCallback {
                    override fun onConnected() {
                        Log.v(TAG, "onConnected")
                        showProgress(false)
                        MyPrefs.userId = userId
                        MyPrefs.company = company
                        MyPrefs.serverUrl = serverUrl
                        startActivity(Intent(this@LoginActivity, MainActivity::class.java))
                        finish()
                    }

                    override fun onFailed(exception: VoicePingException) {
                        Log.v(TAG, "onFailed")
                        showProgress(false)
                        Toast.makeText(
                            this@LoginActivity,
                            R.string.failed_to_sign_in,
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                })
            }
        }
    }

    private fun showProgress(show: Boolean) {
        binding.progressConnect.visibility = if (show) View.VISIBLE else View.GONE
        binding.layoutConnect.visibility = if (show) View.GONE else View.VISIBLE
    }
}